_posts/2018-12-07-fast-starting-clojure-lambdas-using-graalvm.md

```diff
@@ -0,0 +1,80 @@
+
+**Lambda** is a AWS serverless tecnology which lets users to run code without managing servers. Lambdas are paid only for the compute time which is consumed which makes them very attractive option for rarely used web applications. Lambdas itself have a start up time which depends on whether the Lambda application is _cold_ or _warm_. When a Lambda is first time used or there has been about 15 to 30 minutes between the last Lambda usage it takes about 600 ms to start the Lambda. After that Lambda is _warm_ and subsequent invocations are quite similar compared to applications running in normal virtual machines or containers. Performance of a Lambda depends on a memory allocated for it on creation. More memory gives also more computation time.
+
+The start up time is not a problem for Lambdas which are run as cron typed tasks, SQS queue pollers or otherwise where there is no need for quick synchronous responses. 
+
+## Clojure running in JVM AWS Lambdas 
+
+Due to nature of Clojure start up time, using Lambdas have their penalties. In the next table are statistics for different Hello World applications. Clojure applications are tested with 1000 MB and 3000 MB memory for evaluating its effect on Clojure start up.  Other runtimes are tested with 1000 MB memory. Tests were made using a consumer broadband located in Finland against the AWS Ireland region (eu-west 2). The latencies could be better with better network and closer distance to the region. Running tests inside AWS region in EC2 virtual machine would give lower network latencies.
+
+| Memory (MB)| Runtime              | N    | Average    | Standard deviation |
+| ---------- | -------------------- | ---- | ---------- | ------------------ |          
+| 1000       | JVM Java (cold)      | 1000 | 0.585      | 0.108              |
+| 1000       | JVM Java (warm)      | 5000 | 0.205      | 0.063              |
+| 1000       | JVM Clojure (cold)   | 922  | 2.942      | 0.391              |
+| 1000       | JVM Clojure (warm)   | 4600 | 0.237      | 0.133              |
+| 3000       | JVM Clojure (cold)   | 1000 | 2.348      | 0.325              |
+| 3000       | JVM Clojure (warm)   | 5000 | 0.228      | 0.134              |
+| 1000       | Python (cold)        | 1000 | 0.594      | 0.210              |
+| 1000       | Python (warm)        | 5000 | 0.358      | 0.074              |                                                                    
+                                  
+The results are quite problematic for Clojure. Using fastest available Lambda gives a start up time of 2.34 seconds. A start up time using 1000 MB Lambda is near three seconds. The long start up time makes Clojure unusable for example outgoing Slack commands which have a timout of three seconds. Slow start up is recognized in the Clojure community. Solving it may be partially accomplished with changes to Clojure itself. Fortunately new VM has been created with good results for Clojure.  
```

Comment:
```suggestion
The results are quite problematic for Clojure. Using fastest available Lambda gives a start-up time of 2.34 seconds. A start-up time using 1000 MB Lambda is nearly three seconds. The long start-up time makes Clojure unusable for example outgoing Slack commands which have a timeout of three seconds. Slow start-up is recognized in the Clojure community. Solving it may be partially accomplished with changes to Clojure itself. Fortunately, a new VM has been created with good results for Clojure.  
```